File: server/src/main/java/de/ingrid/igeserver/services/MessageService.kt
/**
 * ==================================================
 * Copyright (C) 2023-2024 wemove digital solutions GmbH
 * ==================================================
 * Licensed under the EUPL, Version 1.2 or – as soon they will be
 * approved by the European Commission - subsequent versions of the
 * EUPL (the "Licence");
 *
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */
package de.ingrid.igeserver.services

import de.ingrid.igeserver.persistence.postgresql.jpa.model.ige.Message
import de.ingrid.igeserver.repository.MessageRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.time.OffsetDateTime
import java.util.*

@Service
class MessageService(
    private val messageRepo: MessageRepository,
    private val catalogService: CatalogService
) {

    fun get(catalogId: String?): List<Message> {
        return messageRepo.findAllByCatalog_Identifier(catalogId)
    }

    fun getById(id: Int) : Optional<Message> {
        return messageRepo.findById(id)
    }

    fun getAll(): List<Message> {
        return messageRepo.findAll()
    }

    fun save(
        message:
        de.ingrid.igeserver.model.Message, expires: OffsetDateTime?, catalogId: String?
    ) {
        messageRepo.save(
            Message(
            ).apply {
                this.message = message
                this.expires = expires
                catalog = if (catalogId != null) catalogService.getCatalogById(catalogId) else null
            }
        )

    }

    fun delete(message: Message) {
        messageRepo.delete(message)
    }

}

File: frontend/src/app/services/messages/message.service.ts
/**
 * ==================================================
 * Copyright (C) 2023-2024 wemove digital solutions GmbH
 * ==================================================
 * Licensed under the EUPL, Version 1.2 or – as soon they will be
 * approved by the European Commission - subsequent versions of the
 * EUPL (the "Licence");
 *
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */
import { Injectable } from "@angular/core";
import { BehaviorSubject, Observable } from "rxjs";
import { ConfigService } from "../config/config.service";
import { Message, MessageFormatBackend } from "./message";
import { catchError, tap } from "rxjs/operators";
import { MessagesDataService } from "./messages-data.service";

@Injectable({
  providedIn: "root",
})
export class MessageService {
  messages: Message[] = [];

  messages$ = new BehaviorSubject<Message[]>([]);

  constructor(
    private configService: ConfigService,
    private dataService: MessagesDataService,
  ) {
    this.loadStoredMessages().subscribe((messages) => {
      this.messages = messages;
      this.messages$.next(messages);
    });
  }

  loadStoredMessages(): Observable<Message[]> {
    return this.dataService.loadStoredMessages().pipe(
      tap((storedMessages) =>
        console.debug(`fetched messages`, storedMessages),
      ),
      catchError((e) => {
        const userInfo = this.configService.$userInfo.value;
        console.error("Could not get messages");
        if (userInfo?.assignedCatalogs?.length === 0) {
          console.error("because of user with no assigned catalogs");
          return [];
        } else {
          throw e;
        }
      }),
    );
  }

  getAllMessagesForAdmin(): Observable<MessageFormatBackend[]> {
    return this.dataService.loadAllMessages();
  }

  createMessage(
    message: Message,
    validUntil: Date,
    forCurrentCatalog: boolean,
  ): Observable<any> {
    return this.dataService.createMessage(
      message,
      validUntil,
      forCurrentCatalog,
    );
  }

  deleteMessage(id: Number): Observable<any> {
    return this.dataService.deleteMessage(id);
  }
}
